File: core/flowcatalyst-platform/src/main/java/tech/flowcatalyst/eventtype/EventType.java
package tech.flowcatalyst.eventtype;


import lombok.Builder;
import lombok.With;
import tech.flowcatalyst.platform.shared.TsidGenerator;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents an event type in the FlowCatalyst platform.
 *
 * <p>Event types define the structure and schema for events in the system.
 * Each event type has a globally unique code and can have multiple
 * schema versions for backwards compatibility.
 *
 * <p>Code format: {APPLICATION}:{SUBDOMAIN}:{AGGREGATE}:{EVENT}
 * Example: operant:execution:trip:started
 *
 * <p>Use {@link #create(String, String)} for safe construction with defaults.
 */

@Builder(toBuilder = true)
@With
public record EventType(
    String id,

    /**
     * Unique event type code (globally unique, not tenant-scoped).
     * Format: {application}:{subdomain}:{aggregate}:{event}
     * Example: operant:execution:trip:started
     */
    String code,

    /**
     * Human-friendly name for the event type.
     */
    String name,

    /**
     * Description of the event type.
     */
    String description,

    /**
     * Schema versions for this event type.
     */
    List<SpecVersion> specVersions,

    /**
     * Current status of the event type.
     */
    EventTypeStatus status,

    /**
     * Source of the event type - how it was created (CODE, API/SDK, or UI).
     */
    EventTypeSource source,

    /**
     * Whether events of this type are scoped to a specific client.
     *
     * <p>Client-scoped event types represent events that occur within a client context
     * (e.g., spar:orders:order:created). Non-client-scoped event types are platform-wide
     * (e.g., platform:iam:user:created).
     *
     * <p>This field is immutable after creation. Subscriptions to client-scoped event types
     * must also be client-scoped, and vice versa.
     */
    boolean clientScoped,

    Instant createdAt,

    Instant updatedAt
) {

    /**
     * Find a spec version by version string.
     */
    public SpecVersion findSpecVersion(String version) {
        if (specVersions == null) return null;
        return specVersions.stream()
            .filter(sv -> sv.version().equals(version))
            .findFirst()
            .orElse(null);
    }

    /**
     * Check if all spec versions are in DEPRECATED status.
     */
    public boolean allVersionsDeprecated() {
        if (specVersions == null || specVersions.isEmpty()) {
            return true;
        }
        return specVersions.stream()
            .allMatch(sv -> sv.status() == SpecVersionStatus.DEPRECATED);
    }

    /**
     * Check if all spec versions are in FINALISING status (never finalized).
     */
    public boolean allVersionsFinalising() {
        if (specVersions == null || specVersions.isEmpty()) {
            return true;
        }
        return specVersions.stream()
            .allMatch(sv -> sv.status() == SpecVersionStatus.FINALISING);
    }

    /**
     * Check if a version string already exists.
     */
    public boolean hasVersion(String version) {
        return findSpecVersion(version) != null;
    }

    // ========================================================================
    // Factory Methods
    // ========================================================================

    /**
     * Create a new event type with required fields and sensible defaults.
     *
     * @param code Unique event type code
     * @param name Human-readable name
     * @param clientScoped Whether events of this type are scoped to a client
     * @return A pre-configured builder with defaults set
     */
    public static EventTypeBuilder create(String code, String name, boolean clientScoped) {
        var now = Instant.now();
        return EventType.builder()
            .id(TsidGenerator.generate())
            .code(code)
            .name(name)
            .specVersions(new ArrayList<>())
            .status(EventTypeStatus.CURRENT)
            .source(EventTypeSource.UI)
            .clientScoped(clientScoped)
            .createdAt(now)
            .updatedAt(now);
    }

    /**
     * Create a new event type from SDK/API with required fields and sensible defaults.
     *
     * @param code Unique event type code
     * @param name Human-readable name
     * @param clientScoped Whether events of this type are scoped to a client
     * @return A pre-configured builder with defaults set
     */
    public static EventTypeBuilder createFromApi(String code, String name, boolean clientScoped) {
        var now = Instant.now();
        return EventType.builder()
            .id(TsidGenerator.generate())
            .code(code)
            .name(name)
            .specVersions(new ArrayList<>())
            .status(EventTypeStatus.CURRENT)
            .source(EventTypeSource.API)
            .clientScoped(clientScoped)
            .createdAt(now)
            .updatedAt(now);
    }

    /**
     * Create a new event type from code (platform event types) with required fields and sensible defaults.
     *
     * @param code Unique event type code
     * @param name Human-readable name
     * @param clientScoped Whether events of this type are scoped to a client
     * @return A pre-configured builder with defaults set
     */
    public static EventTypeBuilder createFromCode(String code, String name, boolean clientScoped) {
        var now = Instant.now();
        return EventType.builder()
            .id(TsidGenerator.generate())
            .code(code)
            .name(name)
            .specVersions(new ArrayList<>())
            .status(EventTypeStatus.CURRENT)
            .source(EventTypeSource.CODE)
            .clientScoped(clientScoped)
            .createdAt(now)
            .updatedAt(now);
    }

    /**
     * Add a spec version to this event type.
     *
     * @param specVersion The spec version to add
     * @return A new EventType with the spec version added
     */
    public EventType addSpecVersion(SpecVersion specVersion) {
        var newVersions = new ArrayList<>(specVersions != null ? specVersions : List.of());
        newVersions.add(specVersion);
        return this.toBuilder()
            .specVersions(newVersions)
            .updatedAt(Instant.now())
            .build();
    }
}

File: flowcatalyst-typescript/apps/flowcatalyst/src/platform/domain/event-type/event-type.ts
/**
 * EventType Entity
 *
 * Represents an event type definition in the FlowCatalyst platform.
 * Event type codes follow the format: {application}:{subdomain}:{aggregate}:{event}
 *
 * Each event type can have multiple spec versions (schemas) that define
 * the structure of events published under this type.
 */

import { generate } from "@flowcatalyst/tsid";
import type { SpecVersion } from "./spec-version.js";
import type { EventTypeStatus } from "./event-type-status.js";
import type { EventTypeSource } from "./event-type-source.js";

/**
 * EventType entity.
 */
export interface EventType {
	readonly id: string;
	readonly code: string;
	readonly name: string;
	readonly description: string | null;
	readonly specVersions: SpecVersion[];
	readonly status: EventTypeStatus;
	readonly source: EventTypeSource;
	readonly clientScoped: boolean;
	readonly application: string;
	readonly subdomain: string;
	readonly aggregate: string;
	readonly createdAt: Date;
	readonly updatedAt: Date;
}

/**
 * Input for creating a new EventType.
 */
export type NewEventType = Omit<EventType, "createdAt" | "updatedAt"> & {
	createdAt?: Date;
	updatedAt?: Date;
};

/**
 * Parse code segments from an event type code.
 * Format: {application}:{subdomain}:{aggregate}:{event}
 */
export function parseCodeSegments(code: string): {
	application: string;
	subdomain: string;
	aggregate: string;
	event: string;
} | null {
	const parts = code.split(":");
	if (parts.length !== 4) return null;
	return {
		application: parts[0]!,
		subdomain: parts[1]!,
		aggregate: parts[2]!,
		event: parts[3]!,
	};
}

/**
 * Build an event type code from segments.
 */
export function buildCode(
	application: string,
	subdomain: string,
	aggregate: string,
	event: string,
): string {
	return `${application}:${subdomain}:${aggregate}:${event}`;
}

/**
 * Create a new event type from UI.
 */
export function createEventType(params: {
	application: string;
	subdomain: string;
	aggregate: string;
	event: string;
	name: string;
	description?: string | null;
	clientScoped?: boolean;
}): NewEventType {
	return {
		id: generate("EVENT_TYPE"),
		code: buildCode(
			params.application,
			params.subdomain,
			params.aggregate,
			params.event,
		),
		name: params.name,
		description: params.description ?? null,
		specVersions: [],
		status: "CURRENT",
		source: "UI",
		clientScoped: params.clientScoped ?? false,
		application: params.application,
		subdomain: params.subdomain,
		aggregate: params.aggregate,
	};
}

/**
 * Create a new event type from API/SDK sync.
 */
export function createEventTypeFromApi(params: {
	application: string;
	subdomain: string;
	aggregate: string;
	event: string;
	name: string;
	description?: string | null;
	clientScoped?: boolean;
}): NewEventType {
	return {
		id: generate("EVENT_TYPE"),
		code: buildCode(
			params.application,
			params.subdomain,
			params.aggregate,
			params.event,
		),
		name: params.name,
		description: params.description ?? null,
		specVersions: [],
		status: "CURRENT",
		source: "API",
		clientScoped: params.clientScoped ?? false,
		application: params.application,
		subdomain: params.subdomain,
		aggregate: params.aggregate,
	};
}

/**
 * Update an event type's metadata.
 */
export function updateEventType(
	eventType: EventType,
	updates: Partial<Pick<EventType, "name" | "description">>,
): EventType {
	return {
		...eventType,
		...updates,
		updatedAt: new Date(),
	};
}

/**
 * Add a spec version to an event type.
 */
export function addSpecVersion(
	eventType: EventType,
	specVersion: SpecVersion,
): EventType {
	return {
		...eventType,
		specVersions: [...eventType.specVersions, specVersion],
		updatedAt: new Date(),
	};
}

/**
 * Update a spec version within an event type.
 */
export function updateSpecVersion(
	eventType: EventType,
	version: string,
	updater: (sv: SpecVersion) => SpecVersion,
): EventType {
	return {
		...eventType,
		specVersions: eventType.specVersions.map((sv) =>
			sv.version === version ? updater(sv) : sv,
		),
		updatedAt: new Date(),
	};
}

/**
 * Find a spec version by version string.
 */
export function findSpecVersion(
	eventType: EventType,
	version: string,
): SpecVersion | undefined {
	return eventType.specVersions.find((sv) => sv.version === version);
}

/**
 * Check if all spec versions are deprecated.
 */
export function allVersionsDeprecated(eventType: EventType): boolean {
	return (
		eventType.specVersions.length > 0 &&
		eventType.specVersions.every((sv) => sv.status === "DEPRECATED")
	);
}

/**
 * Check if all spec versions are in FINALISING status.
 */
export function allVersionsFinalising(eventType: EventType): boolean {
	return eventType.specVersions.every((sv) => sv.status === "FINALISING");
}

/**
 * Archive an event type.
 */
export function archiveEventType(eventType: EventType): EventType {
	return {
		...eventType,
		status: "ARCHIVED",
		updatedAt: new Date(),
	};
}
